DietPi-Software | Netdata: Fix uninstall removing root dir and use netdata user

MichaIng opened the pull request:
**Status**: Ready
- [x] Adjust service file
- [x] Run as `netdata` user
- [x] Update to v1.11.1? https://github.com/netdata/netdata/releases
- [x] Changelog
- [x] Apply official systemd unit

**Testing:**
🈯️ VM Stretch (installer, service start, WebUI)
- [x] Jessie does not yet support `ProtectKernelModules=no` and `ProtectKernelTunables=no` and `ExecStartPre=/bin/chown -R netdata:netdata /var/run/netdata` throws "read-only file system" error 🤔. However works fine, after commenting this lines.

**Reference**: https://github.com/Fourdee/DietPi/issues/2336

**Commit list/description**:
+ DietPi-Software | Netdata: Fix uninstall removing root dir
+ DietPi-Software | Netdata: Use netdata user to run service

MichaIng committed:
v6.20

+ DietPi-Software | Netdata: Fix uninstall removing root dir

MichaIng committed:
v6.20

+ DietPi-Software | Netdata: Use netdata user to run service

MichaIng commented:
## Recheck our install method
- Currently we use our own deb package, create the `netdata` user, but run the service as `root`.
- Either we can skip `netdata` user creation or run the service with this user.

### Comparison with other install methods
____
#### Official: https://docs.netdata.cloud/installer/#one-line-installation
- Default systemd unit: https://raw.githubusercontent.com/netdata/netdata/master/system/netdata.service.in
- Service runs as `netdata`, but `app.plugin` child process runs as root. Not sure how this is achieved with above systemd unit.
  - `PermissionsStartOnly=true`
    > Takes a boolean argument. If true, the permission-related execution options, as configured with User= and similar options (see systemd.exec(5) for more information), are only applied to the process started with ExecStart=, and not to the various other ExecStartPre=, ExecStartPost=, ExecReload=, ExecStop=, and ExecStopPost= commands. If false, the setting is applied to all configured commands the same way. Defaults to false.
  - Only applies to:
    ```
    ExecStartPre=/bin/mkdir -p /var/cache/netdata
    ExecStartPre=/bin/chown -R netdata:netdata /var/cache/netdata
    ExecStartPre=/bin/mkdir -p /var/run/netdata
    ExecStartPre=/bin/chown -R netdata:netdata /var/run/netdata
    #ExecStopPost=/bin/rm /var/run/netdata/netdata.pid
    ```
- `netdata` user home dir: `/var/lib/netdata`
- File and dir permissions: https://docs.netdata.cloud/docs/netdata-security/#netdata-directories
  ```
  root@VM-Stretch:~# l /usr/lib/netdata/
  total 4
  drwxr-xr-x 7 root netdata 4096 Dec  9 17:49 conf.d
  root@VM-Stretch:~# l /usr/lib/netdata/conf.d/
  total 96
  -rw-r----- 1 root netdata  9606 Dec  9 17:49 apps_groups.conf
  drwxr-xr-x 2 root netdata  4096 Dec  9 17:49 charts.d
  -rw-r----- 1 root netdata  1829 Dec  9 17:49 charts.d.conf
  -rw-r----- 1 root netdata  1233 Dec  9 17:49 fping.conf
  -rw-r----- 1 root netdata 36724 Dec  9 17:49 health_alarm_notify.conf
  drwxr-xr-x 2 root netdata  4096 Dec  9 17:49 health.d
  -rw-r----- 1 root netdata    57 Dec  9 17:49 health_email_recipients.conf
  drwxr-xr-x 2 root netdata  4096 Dec  9 17:49 node.d
  -rw-r----- 1 root netdata  1310 Dec  9 17:49 node.d.conf
  drwxr-xr-x 2 root netdata  4096 Dec  9 17:49 python.d
  -rw-r----- 1 root netdata  1892 Dec  9 17:49 python.d.conf
  drwxr-xr-x 2 root netdata  4096 Dec  9 17:49 statsd.d
  -rw-r----- 1 root netdata  7564 Dec  9 17:49 stream.conf
  root@VM-Stretch:~# l /usr/libexec/netdata/
  total 16
  drwxr-xr-x 2 root root 4096 Dec  9 17:49 charts.d
  drwxr-xr-x 3 root root 4096 Dec  9 17:49 node.d
  drwxr-xr-x 2 root root 4096 Dec  9 17:49 plugins.d
  drwxr-xr-x 3 root root 4096 Dec  9 17:49 python.d
  root@VM-Stretch:~# l /usr/libexec/netdata/plugins.d/
  total 312
  -rwxr-xr-x 1 root root      154 Dec  9 17:49 alarm-email.sh
  -rwxr-xr-x 1 root root    76509 Dec  9 17:49 alarm-notify.sh
  -rwxr-xr-x 1 root root      301 Dec  9 17:49 alarm-test.sh
  -rwsr-x--- 1 root netdata 97424 Dec  9 17:49 apps.plugin
  -rwxr-xr-x 1 root root     5897 Dec  9 17:49 cgroup-name.sh
  -rwsr-x--- 1 root netdata 39608 Dec  9 17:49 cgroup-network
  -r-xr-x--- 1 root root     6931 Dec  9 17:49 cgroup-network-helper.sh
  -rwxr-xr-x 1 root root     1259 Dec  9 17:49 charts.d.dryrun-helper.sh
  -rwxr-xr-x 1 root root    19123 Dec  9 17:49 charts.d.plugin
  -rwxr-xr-x 1 root root     5516 Dec  9 17:49 fping.plugin
  -rw-r--r-- 1 root root     6631 Dec  9 17:49 loopsleepms.sh.inc
  -rwxr-xr-x 1 root root    11060 Dec  9 17:49 node.d.plugin
  -rwxr-xr-x 1 root root    14781 Dec  9 17:49 python.d.plugin
  -rwxr-xr-x 1 root root     7879 Dec  9 17:49 tc-qos-helper.sh
  ```
____
#### Debian `sid` repo package: https://packages.debian.org/de/sid/netdata
- All dependencies should be available in `stretch` and `stretch-backports`.
  ```
  root@VM-Stretch:~# cat /etc/apt/preferences.d/dietpi-sid
  Package: *
  Pin: release n=sid
  Pin-Priority: -1

  Package: netdata netdata-data
  Pin: release n=sid
  Pin-Priority: 999

  Package: *
  Pin: release n=stretch-backports
  Pin-Priority: 99

  Package: python3-urllib3
  Pin: release n=stretch-backports
  Pin-Priority: 999
  root@VM-Stretch:~# cat /etc/apt/sources.list.d/dietpi-sid.list
  deb https://deb.debian.org/debian/ sid main
  ```
- 🈯️ Jep, with above priorities only `netdata` + `netdata-data` are installed from `sid`, `python3-urllib3` from `stretch-backports`:
  ```
  root@VM-Stretch:~# apt install netdata
  Reading package lists... Done
  Building dependency tree
  Reading state information... Done
  The following additional packages will be installed:
    dh-python fonts-font-awesome libcap2-bin libexpat1 libjs-bootstrap libmpdec2 libpython3-stdlib libpython3.5-minimal libpython3.5-stdlib libyaml-0-2 mime-support netdata-data python3 python3-minimal python3-six python3-urllib3 python3-yaml python3.5 python3.5-minimal
  Suggested packages:
    libdpkg-perl python3-doc python3-tk python3-venv python3-cryptography python3-idna python3-openssl python3-socks python3.5-venv python3.5-doc binutils binfmt-support
  Recommended packages:
    libpam-cap file xz-utils fping nodejs
  The following NEW packages will be installed:
    dh-python fonts-font-awesome libcap2-bin libexpat1 libjs-bootstrap libmpdec2 libpython3-stdlib libpython3.5-minimal libpython3.5-stdlib libyaml-0-2 mime-support netdata netdata-data python3 python3-minimal python3-six python3-urllib3 python3-yaml python3.5 python3.5-minimal
  0 upgraded, 20 newly installed, 0 to remove and 0 not upgraded.
  Need to get 7,716 kB of archives.
  After this operation, 35.6 MB of additional disk space will be used.
  Do you want to continue? [Y/n] y
  Get:1 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libpython3.5-minimal amd64 3.5.3-1+deb9u1 [573 kB]
  Get:2 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libexpat1 amd64 2.2.0-2+deb9u1 [83.4 kB]
  Get:3 https://cdn-aws.deb.debian.org/debian stretch/main amd64 python3.5-minimal amd64 3.5.3-1+deb9u1 [1,691 kB]
  Get:4 https://cdn-aws.deb.debian.org/debian stretch/main amd64 python3-minimal amd64 3.5.3-1 [35.3 kB]
  Get:5 https://cdn-aws.deb.debian.org/debian stretch/main amd64 mime-support all 3.60 [36.7 kB]
  Get:6 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libmpdec2 amd64 2.4.2-1 [85.2 kB]
  Get:7 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libpython3.5-stdlib amd64 3.5.3-1+deb9u1 [2,167 kB]
  Get:8 https://cdn-aws.deb.debian.org/debian stretch/main amd64 python3.5 amd64 3.5.3-1+deb9u1 [229 kB]
  Get:9 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libpython3-stdlib amd64 3.5.3-1 [18.6 kB]
  Get:10 https://cdn-aws.deb.debian.org/debian stretch/main amd64 dh-python all 2.20170125 [86.8 kB]
  Get:11 https://cdn-aws.deb.debian.org/debian stretch/main amd64 python3 amd64 3.5.3-1 [21.6 kB]
  Get:12 https://cdn-aws.deb.debian.org/debian stretch/main amd64 fonts-font-awesome all 4.7.0~dfsg-1 [514 kB]
  Get:13 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libcap2-bin amd64 1:2.25-1 [26.5 kB]
  Get:14 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libjs-bootstrap all 3.3.7+dfsg-2 [274 kB]
  Get:15 https://cdn-aws.deb.debian.org/debian stretch/main amd64 libyaml-0-2 amd64 0.1.7-2 [47.6 kB]
  Get:16 https://cdn-aws.deb.debian.org/debian sid/main amd64 netdata-data all 1.10.0+dfsg-3 [1,044 kB]
  Get:17 https://cdn-aws.deb.debian.org/debian stretch/main amd64 python3-six all 1.10.0-3 [14.4 kB]
  Get:18 https://cdn-aws.deb.debian.org/debian stretch-backports/main amd64 python3-urllib3 all 1.24-1~bpo9+1 [96.9 kB]
  Get:19 https://cdn-aws.deb.debian.org/debian stretch/main amd64 python3-yaml amd64 3.12-1 [108 kB]
  Get:20 https://cdn-aws.deb.debian.org/debian sid/main amd64 netdata amd64 1.10.0+dfsg-3 [563 kB]
  Fetched 7,716 kB in 5s (1,382 kB/s)
  ```
- `nodejs`, as recommendation, is not installed, so it will not conflict with our outside of APT installed version.
- Runs fully as `netdata:netdata`, even the `app.plugin` process. Systemd unit again looks a bid different:
  ```
  root@VM-Stretch:~# cat /lib/systemd/system/netdata.service
  # netdata systemd target

  [Unit]
  Description=netdata - Real-time performance monitoring
  Documentation=man:netdata
  Documentation=file:///usr/share/doc/netdata/html/index.html
  Documentation=https://github.com/firehol/netdata
  After=network-online.target httpd.service squid.service nfs-server.service mysqld.service named.service postfix.service
  ConditionPathExists=/etc/netdata/netdata.conf

  [Service]
  Type=simple
  Environment="netdata_LOG_LOCATION=/var/log/netdata/log"
  ExecStart=/usr/sbin/netdata -D
  ExecReload=/usr/sbin/netdata reload
  TimeoutStopSec=10
  KillMode=mixed
  KillSignal=SIGTERM
  OOMScoreAdjust=-900

  User=netdata
  Group=netdata
  Restart=on-abnormal
  RestartSec=2s
  LimitNOFILE=65536

  WorkingDirectory=/tmp

  # Hardening

  NoNewPrivileges=false
  PermissionsStartOnly=true
  # CAP_SETGID is required for setgroups()
  # CAP_NET_RAW is needed by fping, see #864370
  CapabilityBoundingSet=CAP_DAC_READ_SEARCH CAP_SYS_PTRACE CAP_SETGID CAP_SETUID CAP_NET_RAW
  PrivateTmp=true
  ProtectHome=read-only
  ProtectSystem=full

  ReadOnlyDirectories=/
  ReadWriteDirectories=/proc/self
  ReadWriteDirectories=/var

  # Access to devices and kernel modules and tunables is required
  PrivateDevices=no
  ProtectKernelModules=no
  ProtectKernelTunables=no

  StandardOutput=syslog+console
  StandardError=syslog+console

  [Install]
  WantedBy=multi-user.target
  ```

MichaIng commented:
@Fourdee 
- Do you remember any reason, why we run `netdata` as root, was this required for some features? Might be required for some plugins, since with official install method the plugins process runs as root (see above).
- Debian repo install runs everything as `netdata`, so I guess it is okay, if we follow this? For power users it's not hard to change the systemd unit and on default install I could not find any chart or something on webpage that has issues to gather required infos e.g.
- The official install method includes complex file/dir permissions and several hardening on systemd unit and such. Using official installer includes in-place build, `git` and `nodejs` install, which might conflict with our installs.
- Using our own deb package allows a much faster and more lightweight install, but applying complex permissions and hardening and maintenance increases effort or decreases software security/features.
- Using Debian `sid` package is another alternative that allows lightweight install and reduces maintenance effort for us. But on `sid` things might change fast, e.g. dependencies and such, which might break install method, having `sid` repo in place might have other unwanted side effects and the result does not look much different than with our own deb package...
- Not sure which systemd unit/permission steps from official install method are (or can be) included when creating an own new deb package from source?

Fourdee commented:
@MichaIng 

> Do you remember any reason, why we run netdata as root,

Unsure if i'am honest. I can't for the life of me find the ticket where we changed all software from ```root``` to its own user, info might be there.

> Using Debian sid package is another alternative that allows lightweight install and reduces maintenance effort for us.

Too risky and pre-reqs could be outdated in Stretch at any point in future causing install fail. Opens a can of worms.

> Systemd unit again looks a bid different:

Yep, i'd prefer we base our service on the official one: https://raw.githubusercontent.com/netdata/netdata/master/system/netdata.service.in, however, no harm in adding security hardening.

---

I'll update the binaries to latest version.

We should also add ```CPUSchedulingPolicy=idle``` to our service file. Make sense to default this to idle.

Fourdee committed:
v6.20

+ Netdata 1.11.1 binaries.

MichaIng commented:
It was with v6.12:
- https://github.com/Fourdee/DietPi/blob/master/dietpi/patch_file#L766
- https://github.com/Fourdee/DietPi/issues/1877#issuecomment-403298679
- Netdata not in the list
- `netdata` user already created on v6.2: https://github.com/Fourdee/DietPi/blob/10bd40fc67c3b74c806dc35a8f28ba55844eb5cc/dietpi/dietpi-software#L10393-L10415
- But service was already running as root: https://github.com/Fourdee/DietPi/blob/10bd40fc67c3b74c806dc35a8f28ba55844eb5cc/dietpi/conf/netdata.service
- It think because of user creation, we have simply overseen it. Let's see what the new binary brings about permissions and such and I think we can then safely run it as `netdata`, align with Debian repo install in worst case with issue only on optional plugins.

Fourdee commented:
@MichaIng 

> Let's see what the new binary brings about permissions and such and I think we can then safely run it as netdata, align with Debian repo install in worst case with issue only on optional plugins.

Yep 👍 

Just checked, although set to root, netdata switches to ```netdata``` user during run.
```
root@Fuzon:~# ps aux | grep netdata
netdata  14572  1.0  0.8 174076 28916 ?        Ssl  14:51   0:00 /usr/sbin/netdata -D
netdata  14583  0.1  0.0   9696  2812 ?        S    14:51   0:00 bash /usr/libexec/netdata/plugins.d/tc-qos-helper.sh 1
netdata  14590  0.7  0.0  18820  2720 ?        S    14:51   0:00 /usr/libexec/netdata/plugins.d/apps.plugin 1
netdata  14591  2.6  0.7  92160 24280 ?        Sl   14:51   0:01 /usr/bin/python /usr/libexec/netdata/plugins.d/python.d.plugin 1
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5197,7 +5197,7 @@ _EOF_
 			Banner_Installing
 
-			INSTALL_URL_ADDRESS='https://dietpi.com/downloads/binaries/all/netdata_1.9.0_'
+			INSTALL_URL_ADDRESS='https://dietpi.com/downloads/binaries/all/netdata_1.11.1-1_'
```
@Fourdee 
What do you think about not including version string in the link? That way we can update the binary without the need to touch code, and, as well for older DietPi versions. Also then we don't need to keep all the older binaries on the server just to allow install from older DietPi versions.

Fourdee replied:
@MichaIng 

> What do you think about not including version string in the link? That way we can update the binary without the need to touch code, and, as well for older DietPi versions.

I'd prefer to keep the version codes. EG: if netdata 1.12 requires additional pre-reqs, or, different permissions, it could break DietPi v6.19 or lower installations.

MichaIng replied:
@Fourdee 
Jep just had the same idea while writing... Good point

MichaIng commented:
@Fourdee 
> although set to root, netdata switches to netdata user during run.

Interesting, okay another vote. Although then no harm as well to use official systemd unit. It runs as netdata directly, but has `PermissionsStartOnly=true`, which allows init chown/chmod as root in ExecStartPre.
Only thing, there are many variables inside that are obviously are replaced by installer, e.g. `@localstatedir_POST@`. I will check the outcome and adjust ours then accordingly.

Fourdee commented:
Binaries done and uploaded 🈯️

MichaIng committed:
Merge branch 'dev' into netdata

MichaIng committed:
v6.20

+ CHANGELOG | Netdata update

MichaIng committed:
v6.20

+ DietPi-Patch | Reinstall Netdata to apply new user, permissions and service
+ DietPi-Patch | Cleanup doubled/obsolete reinstalls

MichaIng commented:
### Reinstall cleanup in `patch_file`:
ID | Reinstall on patch to:
-- | ---------------------
6 | 6.19
23 | 6.9
27 | 6.9
28 | 6.9
29 | 6.8
31 | 6.1
32 | 6.12
33 | 6.12
34 | 6.9
35 | 6.5
37 | 6.5
39 | 6.12
45 | 6.20
48 | 6.19
59 | 6.3
65 | 6.20
73 | 6.1
96 | 6.20
106 | 6.18
108 | 6.8
111 | 6.4
113 | 6.19
116 | 6.12
118 | 6.5
119 | 6.9
120 | 6.9
121 | 6.18
123 | 6.18
124 | 6.20
128 | 6.20
129 | 6.18
131 | 6.20
132 | 6.3
134 | 6.1
137 | 6.9
143 | 6.12
144 | 6.18
145 | 6.18
146 | 6.15
147 | 6.18
148 | 6.18
149 | 6.12
154 | 6.12
160 | 6.19
163 | 6.19
166 | 6.8

MichaIng committed:
v6.20

+ DietPi-Software | Netdata: Simplify binary download based on HW arch; Apply permissions according to docs and logs

MichaIng committed:
v6.20

+ DietPi-Software | Netdata: Adjust to match official unit + hardenings from Debian sid repo unit

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5198,32 +5198,8 @@ _EOF_
-
 			DEPS_LIST='zlib1g-dev'
-			Download_Install "$INSTALL_URL_ADDRESS"
+			Download_Install "https://dietpi.com/downloads/binaries/all/netdata_1.11.1-1_$G_HW_ARCH_DESCRIPTION"
```
@Fourdee 
What do you think about it to simplify arch based download links? I already renamed the files on server accordingly.

Fourdee replied:
@MichaIng 

Yep, nice one 👍

MichaIng replied:
At least when updating our binaries, we can apply this as well to other software titles.

MichaIng commented:
@Fourdee 
Okay systemd unit now based on official one + hardenings (and required permissions+capabilities) from Debian sid version. Those make sense and match the official folder permissions from docs as well.

Actually I was hoping, that the capabilities would allow `apps.plugin` to run not as root, since it does not on Debian sid. But I still get the following error log:
```
2018-12-20 16:40:33: apps.plugin ERROR : MAIN : apps.plugin should either run as root (now running with uid 999, euid 999) or have special capabilities. Without these, apps.plugin cannot report disk I/O utilization of other processes. Your system does not support capabilities. To enable setuid to root run: sudo chown root:netdata /usr/libexec/netdata/plugins.d/apps.plugin; sudo chmod 4750 /usr/libexec/netdata/plugins.d/apps.plugin;  (errno 13, Permission denied)
...
2018-12-20 16:40:34: apps.plugin ERROR : MAIN : Cannot process /proc/1/io (command 'systemd') (errno 13, Permission denied)
2018-12-20 16:40:34: apps.plugin ERROR : MAIN : Cannot process entries in /proc/1/fd (command 'systemd') (errno 13, Permission denied)
2018-12-20 16:40:34: apps.plugin ERROR : MAIN : Cannot process /proc/2/io (command 'kthreadd') (errno 13, Permission denied)
2018-12-20 16:40:34: apps.plugin ERROR : MAIN : Cannot process entries in /proc/2/fd (command 'kthreadd') (errno 13, Permission denied)
...
```
> Your system does not support capabilities.

🤔 So this is the reason why the CAPS in the unit have no effect. Reading: https://linux.die.net/man/7/capabilities
> Since kernel 2.5.27, capabilities are an optional kernel component, and can be enabled/disabled via the CONFIG_SECURITY_CAPABILITIES kernel configuration option.
- So this is missing on VM?

However, we leave the CAPS inside the systemd unit, but as well do `chmod 4750 /usr/libexec/netdata/plugins.d/apps.plugin` to be fail safe. The latter also matches the result of official installer, leading to `apps.plugin` running as root, even that the systemd unit runs as netdata.
__________
🈯️ Works well with this (currently tested on Stretch). No unexpected error messages anymore, just many because of missing files (config files => revert to defaults; features not present on system), disabled IPv6 and such, but all expected.
__________
🈯️ Passed my testing, ready to merge

One last concern:
- The web UI offers an updater, which does not clearly show helpful information, but does not work and seems to rely on netdata being cloned via Git.
- For now I think the above is fine, but on the long term actually I would prefer to allow internal updating and since this requires Git anyway, go with official installer. We just need to check, if it works well to pre-install NodeJS and Git our way and then either skip this within installer or check if it does not break it.

MichaIng committed:
v6.20

+ Whoopsie

MichaIng committed:
v6.20

+ DietPi-Software | Netdata: Workaround Jessie not supporting some systemd unit directives

MichaIng committed:
Merge branch 'dev' into netdata

MichaIng committed:
v6.20

+ DietPi-Patch | Resolve conflicts

MichaIng committed:
v6.20

+ DietPi-Patch | Resolve conflicts

MichaIng commented:
@Fourdee 
Will merge to avoid further conflicts. Passed tests here on all distros.
